### DISCUSSION TITLE
Load server SSL/TLS certificates from database (or other dynamic source)

### DISCUSSION THREAD

**Discussion by erik-wramner** [quarkusio/quarkus#33142] [Q&A] [answered]
We have a requirement to be able to change the TLS/SSL keystore and trust stores from the user interface. The trust stores are not a problem, but I can't find a way to get Quarkus to load the TLS keystore from a dynamic source. Ideally, I would like to register a "keystore producer" that can load the keystore with custom code. Then, I could load it from the database on startup. If it changes, the application can stop and rely on Kubernetes for a restart to pick up the new version.

Is there a way to do this that I have missed?

[a comment by an automated account is not shown]

**Reply by erik-wramner** [reply]
I don't seem to be able to edit the labels, but this is not area/kubernetes but rather area/ssl or area/http.

**Reply by Sgitario** [reply]
I've just updated the labels.

**Comment by sberyozkin** [comment]
Can `HttpClientOptionsCustomizer` help ?

**Reply by sberyozkin** [reply]
Sorry, this must be a server options one

**Comment by erik-wramner** [comment]
Issue [30735](https://github.com/quarkusio/quarkus/issues/30735) asks for a similar feature, but is open.

**Comment by sberyozkin** [comment, marked as answer]
@erik-wramner, @cescoffier has added an option to customize Vert.x HTTP configuration with `io.quarkus.vertx.http.HttpServerOptionsCustomizer`, so you can register an `ApplicationScoped` customizer, fetch the keystore data from the database etc (sorry I was not clear about it in the previous comment), can you try it ?
@cescoffier Should we have #30735 resolved ?

**Reply by erik-wramner** [reply]
That is exactly what we need, thanks. I'll give it a try.
